src/pubsub/ua_pubsub_keystorage.h

```diff
@@ -0,0 +1,403 @@
+ *
+ * Each key has a life time, after which the current key is expired and move to
+ * next key in the existing list. For this a callback function is added to the
+ * server. The callback function MoveToNextkey is added to the server as a
+ * timed callback. The addMoveToNextKeyCallback function calculates the time
+ * stamp to trigger the callback when the current Key expires and roll
+ * over to the next key in existing list.
+ *
+ */
+
+/**
+ * @brief This structure holds the information about the keys
+ */
+typedef struct UA_PubSubKeyListItem {
+    /* The SecurityTokenId associated with Key*/
+    UA_UInt32 keyID;
+
+    /* This key is not used directly since the protocol associated with the PubSubGroup(s)
+     * specifies an algorithm to generate distinct keys for different types of
+     * cryptography operations*/
+    UA_ByteString key;
+
+    /* Pointer to the next key data in the list*/
+    struct UA_PubSubKeyListItem *next;
+} UA_PubSubKeyListItem;
+
+/**
+ * @brief This structure holds all info and keys related to one SecurityGroup.
+ * it is used as a list.
+ */
+typedef struct UA_PubSubKeyStorage {
+
+    /**
+     * security group id of the security group related to this storage
+     */
+    UA_String securityGroupID;
+
+    /**
+     * none-owning pointer to the security policy related to this storage
+     */
+    UA_PubSubSecurityPolicy *policy;
+
+    /**
+     * in case of the SKS server, the key storage structure is deleted when removing the
+     * security group.
+     * in case of publisher / subscriber, one key storage structure is
+     * referenced by multiple reader / writer groups have a reference count to manage free
+     */
+    UA_UInt32 referenceCount;
+
+    /**
+     * array of keys. the elements inside this array have a next pointer.
+     * keyList can therefore be used as linked list.
+     */
+    UA_PubSubKeyListItem *keyList;
+
+    /**
+     * size of the keyList
+     */
+    size_t keyListSize;
+
+    /**
+     * The maximum number of Past keys a keystorage is allowed to store
+     */
+    UA_UInt32 maxPastKeyCount;
+
+    /**
+     * The maximum number of Future keys a keyStorage is allowed to store
+     */
+    UA_UInt32 maxFutureKeyCount;
+
+    /*
+     * The maximum keylist size, calculated from maxPastKeyCount and maxFutureKeyCount
+     */
+    UA_UInt32 maxKeyListSize;
+
+    /**
+     * The SecurityTokenId that appears in the header of messages secured with the
+     * CurrentKey. It starts at 1 and is incremented by 1 each time the KeyLifetime
+     * elapses even if no keys are requested. If the CurrentTokenId increments past the
+     * maximum value of UInt32 it restarts a 1.
+     */
+    UA_UInt32 currentTokenId;
+
+    /**
+     *  the current key used to secure the messages
+     */
+    UA_PubSubKeyListItem *currentItem;
+
+    /**
+     * the last key item in the storage
+     */
+    UA_PubSubKeyListItem *lastItem;
```

Comment:
If you need a list with first and last element known, use the TAILQ structure from /deps/open61541_queue.h.